pyomo/solver/ipopt.py

```diff
@@ -0,0 +1,517 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2022
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import os
+import subprocess
+import datetime
+import io
+import sys
+from typing import Mapping, Optional, Dict
+
+from pyomo.common import Executable
+from pyomo.common.config import ConfigValue, NonNegativeInt, NonNegativeFloat
+from pyomo.common.errors import PyomoException
+from pyomo.common.tempfiles import TempfileManager
+from pyomo.core.base import Objective
+from pyomo.core.base.label import NumericLabeler
+from pyomo.core.staleflag import StaleFlagManager
+from pyomo.repn.plugins.nl_writer import NLWriter, NLWriterInfo, AMPLRepn
+from pyomo.solver.base import SolverBase, SymbolMap
+from pyomo.solver.config import SolverConfig
+from pyomo.solver.factory import SolverFactory
+from pyomo.solver.results import (
+    Results,
+    TerminationCondition,
+    SolutionStatus,
+    parse_sol_file,
+)
+from pyomo.solver.solution import SolutionLoaderBase, SolutionLoader
+from pyomo.common.tee import TeeStream
+from pyomo.common.log import LogStream
+from pyomo.core.expr.visitor import replace_expressions
+from pyomo.core.expr.numvalue import value
+from pyomo.core.base.suffix import Suffix
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class ipoptSolverError(PyomoException):
+    """
+    General exception to catch solver system errors
+    """
+
+
+class ipoptConfig(SolverConfig):
+    def __init__(
+        self,
+        description=None,
+        doc=None,
+        implicit=False,
+        implicit_domain=None,
+        visibility=0,
+    ):
+        super().__init__(
+            description=description,
+            doc=doc,
+            implicit=implicit,
+            implicit_domain=implicit_domain,
+            visibility=visibility,
+        )
+
+        self.executable = self.declare(
+            'executable', ConfigValue(default=Executable('ipopt'))
+        )
+        self.save_solver_io: bool = self.declare(
+            'save_solver_io', ConfigValue(domain=bool, default=False)
+        )
+        # TODO: Add in a deprecation here for keepfiles
+        self.temp_dir: str = self.declare(
```

Comment:
Might make more sense to call this `working_dir` since it's really the directory in which we want to create/whatever files.